src/util.ts

```diff
@@ -0,0 +1,22 @@
+/**
+ * Used for functions that take an options objects. Fills in defaults for
+ * options not provided.
+ *
+ * @internal
+ */
+export const applyDefaults = <T extends Record<string, any>>(
```

Comment:
I was fighting a bug for a while where the "provided" value had field present but set to undefined. This caused the default value to be overridden with undefined. This utility is safer than Object.assign or spread operators for applying defaults in case the provided value contains any undefined fields.

Reply:
Sounds a like a user-level bug.

Basically this function ingores undefined values.

There may be a scenario in the future where a user wants to override some default setting to undefined  - imagine if we had a new client config control that was on by default but could be turned on by setting to undefined.

IMO we should keep using the spread operators. That way the interface is simple - you get what you ask for. If you ask for undefined on a value that's what you get.

If the bug was hard to track down we can add some client config validation to catch those immediately and give a clean error to the user.

Reply:
There is MUCH to be said about keeping it simple! And I sometimes fall in to a trap of making things more complicated than they need to be, so this is always good to be checked on 🙂. Here's some background on this one:

I created the bug within the driver by defining one of the timeouts as possibly undefined. That explicit undefined value was passed down through the HTTPRequest. At the time, the default was being applied/enforced in the HTTPClient and was getting overridden by that explicit undefined. I noticed the potential for this issue at client configuration, too, so created the reusable util based on the [v4 driver](https://github.com/fauna/faunadb-js/blob/3940ad9e71907b7795fd975ae16c9392a679b053/src/_util.js#L348-L365). (We have the benefit of ES6 and typescript to simplify the method and enforce mandatory options without setting undefined in the default)

This is not relevant to the HTTPCient anymore (timeouts and session idle time are defined before passing to the http client), but I think it is still important to apply defaults this way. That is, important to not let explicit undefined options override the defaults. We rely on the defaults being present, and now we have more of them.

```typescript
export const DEFAULT_CLIENT_CONFIG: Omit<ClientConfiguration, "secret"> = {
  client_timeout_buffer_ms: 500,
  endpoint: endpoints.default,
  format: "tagged",
  http2_session_idle_ms: 500,
  max_conns: 10,
  query_timeout_ms: 5000
};
```

We added `client_timeout_buffer_ms`, `http2_session_idle_ms`, and `query_timeout_ms` to mandatory client configuration. The driver assumes these values are defined, so will blow up they are set to undefined.